### PR TITLE
Improve DTD/XSD security with regard to remote resources

### LINKED ISSUE #671: NPE in ContentModelCodeActionParticipant.doCodeAction#L47
I'm not able to reproduce, so I don't understand how it occured, but I saw some NPEs in https://github.com/eclipse/lemminx/blob/f12cf5d1cbd47f59bdb4b632bed57b11b8195be2/org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/ContentModelCodeActionParticipant.java#L47

### PR DESCRIPTION
**Pull request by angelozerr**
Improve DTD/XSD security with regard to remote resources

Fixes [#671](https://github.com/redhat-developer/vscode-xml/issues/671)

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
This PR fixes following problem with our DTD validator:

 * `higlight the SYSTEM url `which have problem
 * use `xml.validation.resolveExternalEntities` used to enable/disable the resolved of external entities.
 * use `xml.downloadExternalResources.enabled` to allow/forbid download of SYSTEM url inside DTD.

Here a demo with this draft PR:

DTDDownloadSupport

**Comment by angelozerr** [comment]
@fbricon you can start playing with the PR. All tests are working now.

I need to check if XSD is working too (when you reference xsd with xs:import)

**Comment by angelozerr** [comment]
> I need to check if XSD is working too (when you reference xsd with xs:import)

It should be worked with XSD. The only thing that it doesn't work is the `Force download with XSD`. It's a problem with XSD Xerces. My validation arguments  (which contains the uri to force to download)  with XSD is lost when I consume the CacheResourceManager. I don't know how to manage that except with a ThreadLocal.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/commands/XMLValidationFileCommand.java [hunk 1]
```diff
@@ -52,12 +57,16 @@ public XMLValidationFileCommand(ContentModelManager contentModelManager, IXMLDoc
...
+		// Validation args can contains the external resource ('url' as key map) to
+		// force do download.
+		JsonObject validationArgs = params.getArguments().size() > 1 ? (JsonObject) params.getArguments().get(1) : null;
 		// 1. remove the referenced grammar in the XML file from the Xerces grammar pool
 		// (used by the Xerces validation) and the content model documents cache (used
 		// by the XML completion/hover based on the grammar)
 		contentModelManager.evictCacheFor(document);
 		// 2. trigger the validation for the given XML file
-		validationService.validate(document);
+		Map map = JSONUtility.toModel(validationArgs, Map.class);
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/commands/XMLValidationFileCommand.java:68]: what happens when validationArgs == null?

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/model/ContentModelProvider.java [hunk 2]
```diff
@@ -63,19 +63,58 @@ public String getKind() {
...
+	 * Create content model document (XSD, DTD, etc) from the given resource key and
+	 * null otherwise.
+	 * 
+	 * @param key                     the resource key.
+	 * @param resolveExternalEntities true if external entities can be resolved and
+	 *                                false otherwise.
+	 * @return the content model document (XSD, DTD, etc) from the given resource
+	 *         key and null otherwise.
+	 */
+	CMDocument createCMDocument(String key, boolean resolveExternalEntities);
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/model/ContentModelProvider.java:107]: resolveExternalEntities or downloadExternalResources?

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/commands/XMLValidationFileCommand.java (hunk 1 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/commands/XMLValidationFileCommand.java:68]: No force download url will be done when validation will occurs. The null arguments is managed at https://github.com/eclipse/lemminx/pull/1183/files#diff-07b75b69658ecc851282ac49a9a36f30a26fcb8e26004e6705d279dc094af2ebR91

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/DownloadResourceDisabledCodeAction.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+import com.google.gson.JsonObject;
+
+/**
+ * Code action to force download of an external resource.
+ * 
+ * @author Angelo ZERR
+ *
+ */
+public class DownloadResourceDisabledCodeAction implements ICodeActionParticipant {
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/DownloadResourceDisabledCodeAction.java:41]: DownloadDisabledResourceCodeAction

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/DownloadResourceDisabledCodeAction.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+			String documentURI = document.getDocumentURI();
+
+			Command command = createDownloadCommand(title, url, documentURI);
+
+			codeAction.setCommand(command);
+			codeAction.setDiagnostics(Arrays.asList(diagnostic));
+			codeActions.add(codeAction);
+		} catch (Exception e) {
+			LOGGER.log(Level.SEVERE, "Error while creating download resource disabled code action.", e);
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/DownloadResourceDisabledCodeAction.java:70]: disabled resource

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/dtd/participants/diagnostics/LSPXML11DTDProcessor.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import org.apache.xerces.impl.msg.XMLMessageFormatter;
+import org.apache.xerces.xni.parser.XMLEntityResolver;
+import org.eclipse.lemminx.extensions.xerces.LSPXMLEntityManager;
+
+/**
+ * The Xerces {@link XML11DTDProcessor} doesn't give the capability to
+ * customize the XML entity manager.
+ * 
+ * <p>
+ * This class extends {@link XML11DTDProcessor} and customize the XML entity
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/dtd/participants/diagnostics/LSPXML11DTDProcessor.java:27]: customizes

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/dtd/participants/diagnostics/LSPXML11DTDProcessor.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+ * manager with the {@link LSPXMLEntityManager} which takes care of download
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/dtd/participants/diagnostics/LSPXML11DTDProcessor.java:28]: which handles the "download external entities" setting.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xerces/LSPSchemaParsingConfig.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import org.apache.xerces.impl.XMLEntityManager;
+import org.apache.xerces.impl.XMLErrorReporter;
+import org.apache.xerces.impl.xs.opti.SchemaParsingConfig;
+
+/**
+ * The Xerces {@link SchemaParsingConfig} doesn't give the capability to
+ * customize the XML entity manager.
+ * 
+ * <p>
+ * This class extends {@link SchemaParsingConfig} and customize the XML entity
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xerces/LSPSchemaParsingConfig.java:26]: customizes

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xerces/LSPSchemaParsingConfig.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+ * manager with the {@link LSPXMLEntityManager} which takes care of download
```
Comment (fbricon) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xerces/LSPSchemaParsingConfig.java:27]: see 2 comments above

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/model/ContentModelProvider.java (hunk 2 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/model/ContentModelProvider.java:107]: - resolveExternalEntities is used to configure the Xerces SAX parser when any XML, DTD, XSD is loaded. The resolveExternalEntities is a feature that Xerces supports.
 - downloadRemoteResources is managed on CacheResourcesManager side when an URL must be downloaded. Its XMLCacheResolverExtension an implementation of Xerces XMLEntityResolver which provides the capability to returns the content of an URI.

In other words, we need only resolveExternalEntities to configure Xerces SAX parser.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/DownloadResourceDisabledCodeAction.java (hunk 3 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/DownloadResourceDisabledCodeAction.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/DownloadResourceDisabledCodeAction.java (hunk 4 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/contentmodel/participants/codeactions/DownloadResourceDisabledCodeAction.java]: fixed

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/dtd/participants/diagnostics/LSPXML11DTDProcessor.java (hunk 6 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/dtd/participants/diagnostics/LSPXML11DTDProcessor.java:28]: My comment was wrong because it manages any error when download is processed. So I think 

`any remote resource download errors.`

should be better, what do you think about that?

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xerces/LSPSchemaParsingConfig.java (hunk 7 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xerces/LSPSchemaParsingConfig.java]: fixed

**Review by AlexXuChen — COMMENTED** [review]
It seems that no validation appears when `xml.server.preferBinary` is enabled:
`xml.server.preferBinary` is false
Screenshot from 2022-03-28 11-18-15
`xml.server.preferBinary` is true
Screenshot from 2022-03-28 11-18-50

**Review by AlexXuChen — COMMENTED** [review]
Everything using the Java server is working for me

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xerces/LSPSchemaParsingConfig.java (hunk 8 above)
Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/extensions/xerces/LSPSchemaParsingConfig.java]: fixed

**Comment by angelozerr** [comment]
> It seems that no validation appears when xml.server.preferBinary is enabled:

@rgrunber can you confirm please that it works with binary in Linux please. I tested with Windows and ot works good.

**Comment by rgrunber** [comment]
As mentioned, this works for me and I can be merged. I found an odd issue around the code actions when the schema reference URL contains a query string, but the change itself still worked in restricting the download as needed.

**Comment by angelozerr** [comment]
Thank a lot @fbricon @AlexXuChen @rgrunber for your review!
